designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,331 @@
+
+Every Debug Configuration that references a SAM Template and resource pairing will produce a CodeLens above that resource. When clicked, these CodeLenses launch the corresponding debug session, as if the user selected that debug configuration and pressed Debug from VS Code's Debug view.
+
+An additional CodeLens is placed above every template resource of type `AWS::Serverless::Function` called "Add Debug Configuration". Clicking this CodeLens produces a pre-filled `aws-sam` Debug Configuration in `launch.json` capable of performing a [local invoke](#sam-template-resource-local) of the associated resource.
+
+#### CodeLenses in Code files
+
+CodeLenses in code files provide support for debugging standalone Lambda function handlers.
+
+The following CodeLenses appear over any function that appears to be an eligible Lambda handler: Run Locally, Debug Locally, Configure.
+
+The Configure CodeLens allows users to customize how the function handler is locally debugged. Configuration is optional, using sensible defaults when necessary. The configurable aspects are listed [above](#debug-config). Each handler is configured separately, and all handler configurations are stored in the workspace at `.aws/lambda-handlers.json` ([file structure](#code-file-codelens-config)). Clicking the Configure CodeLens opens this file, adds a configuration entry for the corresponding handler if necessary, and places the cursor at the handler's entry. Users have autocompletion support in this file. A rich UI for configuration is not considered at this time, but the door remains open to adding a visual editor in the future based on user feedback.
+
+When clicked, the Run and Debug CodeLenses locally invoke their associated Lambda handler function. These Lambda handlers are invoked independently of SAM Templates that exist in the users workspace. The Toolkit performs the following:
+
+-   A temporary SAM Template is produced, containing one resource that references the Lambda handler
+-   The temporary SAM Application is built. Where applicable, settings are applied from the configurations speficied with the Configure CodeLens.
+-   The resource in the temporary SAM Template is invoked. Where applicable, settings are applied from the configurations speficied with the Configure CodeLens.
+-   (If the Debug CodeLens was clicked) The VS Code debugger is attached to the invoked resource
+-   SAM CLI output is shown in the Toolkit's Output Channel. If a debugger is attached, output is also shown in the Debug Console.
+
+These CodeLenses do not support API Gateway style invokes.
+
+Some users may find CodeLenses within code files distracting, particularly if they are using the Toolkit for features not related to local debugging. Toolkit settings can be used to enable and disable CodeLenses. CodeLenses only appear for languages/runtimes that the Toolkit has implemented Debug support for.
```

Comment:
Will there be any other access points to this functionality if the codelenses are disabled? This implies that wanting a lack of distraction is mutually exclusive from wanting to debug SAM apps

Reply:
"if they are using the Toolkit for features not related to local debugging"

There are no other access points for running standalone lambda function handlers

Reply:
Are there any other ways to surface this? Could we provide this either in a tooltip when hovering over a handler, right-clicking a handler, or through a sidebar decoration (a la Jetbrains)?